File: src/sc_neurocore/accel/julia/neurons/energy_lif.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia Fardet-Levina eLIF

"""Coupled RK4 author-Brian implementation of Fardet-Levina eLIF."""
module EnergyLifAccel

export EnergyLIFNeuronState, candidate, simulate, step!, valid

"""Complete source eLIF state and configuration."""
mutable struct EnergyLIFNeuronState
    v::Float64; epsilon::Float64; capacitance::Float64; g_leak::Float64
    e_0::Float64; e_u::Float64; e_d::Float64; e_f::Float64
    v_threshold::Float64; v_reset::Float64; alpha::Float64; epsilon_0::Float64
    epsilon_c::Float64; delta::Float64; tau_e::Float64; dt::Float64
end

EnergyLIFNeuronState() = EnergyLIFNeuronState(-61.0, 0.32, 100.0, 9.0, -62.5, -58.5, -40.0, -62.0, -59.0, -62.0, 1.0, 0.5, 0.18, 0.01, 200.0, 0.1)

function rhs(s::EnergyLIFNeuronState, v::Float64, epsilon::Float64, current::Float64)
    leak = s.e_0 + (s.e_u - s.e_0) * (1.0 - epsilon / s.epsilon_0)
    dv = (s.g_leak * (leak - v) + current) / s.capacitance
    de = ((1.0 - epsilon / (s.alpha * s.epsilon_0))^3 - (v - s.e_f) / (s.e_d - s.e_f)) / s.tau_e
    return dv, de
end

"""Return one simultaneous source RK4 candidate."""
function candidate(s::EnergyLIFNeuronState, current::Float64)
    dt = s.dt
    k1 = rhs(s, s.v, s.epsilon, current)
    k2 = rhs(s, s.v + dt * k1[1] / 2, s.epsilon + dt * k1[2] / 2, current)
    k3 = rhs(s, s.v + dt * k2[1] / 2, s.epsilon + dt * k2[2] / 2, current)
    k4 = rhs(s, s.v + dt * k3[1], s.epsilon + dt * k3[2], current)
    return s.v + dt * (k1[1] + 2k2[1] + 2k3[1] + k4[1]) / 6,
           s.epsilon + dt * (k1[2] + 2k2[2] + 2k3[2] + k4[2]) / 6
end

"""Return whether the complete source state is valid."""
function valid(s::EnergyLIFNeuronState)
    values = (s.v, s.epsilon, s.capacitance, s.g_leak, s.e_0, s.e_u, s.e_d, s.e_f,
              s.v_threshold, s.v_reset, s.alpha, s.epsilon_0, s.epsilon_c, s.delta, s.tau_e, s.dt)
    return all(isfinite, values) && -200 <= s.v <= 100 && -200 <= s.v_reset <= 100 &&
           0 <= s.epsilon <= 5 && s.capacitance > 0 && s.g_leak > 0 && s.alpha > 0 &&
           s.epsilon_0 > 0 && s.epsilon_c >= 0 && s.delta >= 0 && s.tau_e > 0 &&
           0 < s.dt <= min(1.0, s.tau_e) && s.e_d != s.e_f && s.v_threshold > s.v_reset
end

"""Advance one source sample, returning `-1` without mutation on failure."""
function step!(s::EnergyLIFNeuronState, current::Float64=0.0)
    if !valid(s) || !isfinite(current); return -1; end
    v, epsilon = candidate(s, current)
    if !(isfinite(v) && -200 <= v <= 100 && isfinite(epsilon) && 0 <= epsilon <= 5); return -1; end
    if v > s.v_threshold && epsilon > s.epsilon_c
        after = epsilon - s.delta
        if !(0 <= after <= 5); return -1; end
        s.v, s.epsilon = s.v_reset, after
        return 1
    end
    s.v, s.epsilon = v, epsilon
    return 0
end

"""Simulate a complete current trace from an explicitly supplied state."""
function simulate(currents::AbstractVector{<:Real}; state::EnergyLIFNeuronState=EnergyLIFNeuronState())
    voltages = zeros(length(currents)); epsilon = zeros(length(currents)); events = zeros(Int, length(currents))
    for i in eachindex(currents)
        events[i] = step!(state, Float64(currents[i]))
        voltages[i] = state.v
        epsilon[i] = state.epsilon
    end
    return (; voltages, epsilon, events, state)
end

end

File: src/sc_neurocore/accel/julia/neurons/energy_lif.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — dedicated Julia neuron facade

from __future__ import annotations

from typing import Any

import numpy as np
import numpy.typing as npt

from ._runtime import JULIA_MAIN as _jl
from ._runtime import KERNEL_DIR as _KERNEL_DIR

_ENERGY_LIF_LOADED = False
_SC_NORMALIZED_ENERGY_LIF_LOADED = False

def _ensure_energy_lif_loaded() -> Any:
    """Include the source-faithful Fardet-Levina eLIF module."""
    global _ENERGY_LIF_LOADED
    if _jl is None:
        raise ImportError("juliacall not available; install the `julia` extra")
    if not _ENERGY_LIF_LOADED:
        jl_path = _KERNEL_DIR / "energy_lif.jl"
        if not jl_path.is_file():
            raise FileNotFoundError(f"energy_lif.jl missing at {jl_path}")
        _jl.include(str(jl_path))
        _ENERGY_LIF_LOADED = True
    return _jl.EnergyLifAccel


def simulate_energy_lif(
    currents: npt.ArrayLike,
    *,
    v: float = -61.0,
    epsilon: float = 0.32,
    capacitance: float = 100.0,
    g_leak: float = 9.0,
    e_0: float = -62.5,
    e_u: float = -58.5,
    e_d: float = -40.0,
    e_f: float = -62.0,
    v_threshold: float = -59.0,
    v_reset: float = -62.0,
    alpha: float = 1.0,
    epsilon_0: float = 0.5,
    epsilon_c: float = 0.18,
    delta: float = 0.01,
    tau_e: float = 200.0,
    dt: float = 0.1,
) -> dict[str, object]:
    """Run the complete source-faithful eLIF trace in Julia."""
    drive = np.ascontiguousarray(currents, dtype=np.float64)
    if drive.ndim != 1 or not np.isfinite(drive).all():
        raise ValueError("Julia EnergyLIF current must be finite and one-dimensional")
    module = _ensure_energy_lif_loaded()
    state = module.EnergyLIFNeuronState(
        v,
        epsilon,
        capacitance,
        g_leak,
        e_0,
        e_u,
        e_d,
        e_f,
        v_threshold,
        v_reset,
        alpha,
        epsilon_0,
        epsilon_c,
        delta,
        tau_e,
        dt,
    )
    result = module.simulate(drive, state=state)
    return {
        "voltages": np.asarray(result.voltages, dtype=np.float64),
        "epsilon": np.asarray(result.epsilon, dtype=np.float64),
        "events": np.asarray(result.events, dtype=np.int64),
        "v_final": float(result.state.v),
        "epsilon_final": float(result.state.epsilon),
    }


def _ensure_sc_normalized_energy_lif_loaded() -> Any:
    """Include the retained normalized-energy SC module."""
    global _SC_NORMALIZED_ENERGY_LIF_LOADED
    if _jl is None:
        raise ImportError("juliacall not available; install the `julia` extra")
    if not _SC_NORMALIZED_ENERGY_LIF_LOADED:
        jl_path = _KERNEL_DIR / "sc_normalized_energy_lif.jl"
        if not jl_path.is_file():
            raise FileNotFoundError(f"sc_normalized_energy_lif.jl missing at {jl_path}")
        _jl.include(str(jl_path))
        _SC_NORMALIZED_ENERGY_LIF_LOADED = True
    return _jl.SCNormalizedEnergyLifAccel


def simulate_sc_normalized_energy_lif(
    currents: npt.ArrayLike,
    *,
    v: float = -70.0,
    epsilon: float = 1.0,
    v_rest: float = -70.0,
    v_reset: float = -70.0,
    v_threshold: float = -50.0,
    tau_m: float = 10.0,
    tau_e: float = 500.0,
    alpha: float = 0.1,
    epsilon_0: float = 1.0,
    resistance: float = 1.0,
    dt: float = 1.0,
) -> dict[str, object]:
    """Run the complete retained normalized-energy SC trace in Julia."""
    drive = np.ascontiguousarray(currents, dtype=np.float64)
    if drive.ndim != 1 or not np.isfinite(drive).all():
        raise ValueError("Julia SC EnergyLIF current must be finite and one-dimensional")
    module = _ensure_sc_normalized_energy_lif_loaded()
    state = module.SCNormalizedEnergyLIFNeuronState(
        v,
        epsilon,
        v_rest,
        v_reset,
        v_threshold,
        tau_m,
        tau_e,
        alpha,
        epsilon_0,
        resistance,
        dt,
    )
    result = module.simulate(drive, state=state)
    return {
        "voltages": np.asarray(result.voltages, dtype=np.float64),
        "epsilon": np.asarray(result.epsilon, dtype=np.float64),
        "events": np.asarray(result.events, dtype=np.int64),
        "v_final": float(result.state.v),
        "epsilon_final": float(result.state.epsilon),
    }
